include/pybind11/cast.h

```diff
@@ -617,122 +634,140 @@ template <> class type_caster<bool> {
 
-    static PYBIND11_DESCR name() { return type_descr(_(PYBIND11_STRING_NAME)); }
-};
+    bool load_codepoint(char32_t codepoint) {
+        if (codepoint > max)
+            return false;
+        value = std::is_signed<CharT>::value
+            ? static_cast<CharT>((typename std::make_unsigned<CharT>::type) codepoint)
+            : (CharT) codepoint;
+        return true;
+    }
 
-template <> class type_caster<wchar_t> : public type_caster<std::wstring> {
 public:
-    bool load(handle src, bool convert) {
-        if (src.is_none()) return true;
-        return type_caster<std::wstring>::load(src, convert);
-    }
 
-    static handle cast(const wchar_t *src, return_value_policy /* policy */, handle /* parent */) {
-        if (src == nullptr) return none().inc_ref();
-        return PyUnicode_FromWideChar(src, (ssize_t) wcslen(src));
+#if PY_VERSION_HEX >= 0x03030000
+    bool load(handle src, bool) {
+        if (!src)
+            return false;
+        if (!PyUnicode_Check(src.ptr()))
+            return false;
+        int ready = PyUnicode_READY(src.ptr());
+        if (ready == -1) { PyErr_Clear(); return false; }
+        if (PyUnicode_GET_LENGTH(src.ptr()) != 1)
+            return false;
+        return load_codepoint(PyUnicode_READ_CHAR(src.ptr(), 0));
     }
-
-    static handle cast(wchar_t src, return_value_policy /* policy */, handle /* parent */) {
-        wchar_t wstr[2] = { src, L'\0' };
-        return PyUnicode_FromWideChar(wstr, 1);
+#else
+    // Python before PEP 393 (implemented in Python 3.3) does not give any particularly nice way to
+    // interact with unicode *characters* (like the above code), so we go through conversion to a
+    // UTF-32 string from which we are guaranteed to be able to read codepoints (otherwise all we
+    // can get out of Python is the number of code units (which is really not helpful in determining
+    // whether we have a single character or not).
+    bool load(handle src, bool convert) {
+        make_caster<std::u32string> strcaster;
+        if (!strcaster.load(src, convert))
+            return false;
+        auto &str = strcaster.operator std::u32string&();
+        if (str.size() != 1)
+            return false;
+        return load_codepoint(str[0]);
     }
+#endif
 
-    operator wchar_t*() { return success ? (wchar_t *) value.c_str() : nullptr; }
-    operator wchar_t&() { return value[0]; }
+    static handle cast(CharT src, return_value_policy, handle) {
+        char32_t codepoint = static_cast<typename std::make_unsigned<CharT>::type>(src);
+        return PyUnicode_Decode(reinterpret_cast<char*>(&codepoint), 4, "utf32", nullptr);
+    }
 
-    static PYBIND11_DESCR name() { return type_descr(_(PYBIND11_STRING_NAME)); }
+    // chr()/unichr() isn't technically a type, but it should get the point across:
+    PYBIND11_TYPE_CASTER(CharT, _(PYBIND11_CHR_NAME "(") + _<(max <= 0xffff)>(_("<=") + _<max>(), _("")) + _(")"));
```

Comment:
I agree that it's a bit strange, but I have a feeling it's going to be rather confusing to see an error about invalid function arguments with a `unichar` accepted, but only certain `unichar` actually accepted.  (I guess this is more or less the same issue with #651).

I suppose, bigger picture, it would be nice to be able to override or supplement the call failure message so that a caster could return an argument failure with a reason (e.g. "Only char values <= 255 are accepted") and have that reason displayed in the exception (but, I suppose, only if the function isn't overloaded).